File: frontend/src/types/seller.ts
import type { RestaurantMenuItem } from "@/types/restaurant";

export type SellerActivityType = "shop" | "restaurant" | "hotel" | "enterprise";
export type StorefrontTier = "basic" | "premium";

export type SellerServiceOffering = {
  title: string;
  description: string | null;
  display_mode: "consult_only" | "book_only" | "shop";
};

export type HotelRoomType = {
  id?: string | null;
  name: string;
  description: string | null;
  night_price: number;
  capacity: number;
  amenities: string[];
  photo_urls: string[];
  deposit_amount: number | null;
};

export type SellerProfile = {
  id: string;
  user_id: string;
  vendor_id: string;
  business_name: string;
  phone: string | null;
  city: string;
  address: string | null;
  activity_type: SellerActivityType;
  storefront_tier: StorefrontTier;
  description: string | null;
  logo_url: string | null;
  cover_image_url: string | null;
  opening_hours: string | null;
  whatsapp_contact: string | null;
  contact_email: string | null;
  gallery_images: string[];
  service_offerings: SellerServiceOffering[];
  room_types: HotelRoomType[];
  deposit_payment_method: "nita" | "amana" | null;
  deposit_amount: number | null;
  commission_rate_override: number | null;
  service_fee_override: number | null;
  seller_subscription_fee_override: number | null;
  onboarding_fee_paid_at: string | null;
  subscription_paid_until: string | null;
  subscription_last_payment_reference: string | null;
  effective_commission_rate: number;
  effective_service_fee: number;
  effective_seller_subscription_fee: number;
  accepts_table_reservations: boolean;
  accepts_hotel_bookings: boolean;
  is_verified: boolean;
  created_at: string;
};

export type SellerSubscriptionStatus = {
  has_seller_profile: boolean;
  onboarding_fee_paid: boolean;
  onboarding_fee_paid_at: string | null;
  subscription_paid_until: string | null;
  subscription_active: boolean;
  monthly_fee: number;
  onboarding_fee: number;
  amount_due_now: number;
  currency: "XOF";
  has_pending_payment_request: boolean;
};

export type SellerSubscriptionPaymentRequest = {
  id: string;
  seller_profile_id: string;
  seller_user_id: string;
  payment_mode: "nita" | "amana";
  transaction_reference: string;
  months: number;
  amount_claimed: number;
  status: "pending" | "approved" | "rejected";
  admin_note: string | null;
  reviewed_by_user_id: string | null;
  reviewed_at: string | null;
  created_at: string;
};

export type SellerProfilePayload = {
  business_name: string;
  phone?: string;
  city: string;
  address?: string;
  activity_type?: SellerActivityType;
  storefront_tier?: StorefrontTier;
  description?: string;
  logo_url?: string;
  cover_image_url?: string;
  opening_hours?: string;
  whatsapp_contact?: string;
  contact_email?: string;
  gallery_images?: string[];
  service_offerings?: SellerServiceOffering[];
  room_types?: HotelRoomType[];
  deposit_payment_method?: "nita" | "amana";
  deposit_amount?: number;
  accepts_table_reservations?: boolean;
  accepts_hotel_bookings?: boolean;
};

export type SellerProductPayload = {
  name: string;
  brand: string;
  description?: string;
  main_image_url?: string;
  category_id?: string;
  amount: number;
  currency: string;
  stock_quantity: number;
};

export type SellerInventoryItem = {
  price_id: string;
  product_id: string;
  product_name: string;
  brand: string;
  amount: number;
  currency: string;
  stock_quantity: number;
  is_active: boolean;
  is_boosted: boolean;
  promo_price: number | null;
  promo_until: string | null;
  boost_until: string | null;
};

export type SellerShopOrderItem = {
  id: string;
  product_id: string;
  product_name: string;
  quantity: number;
  unit_price: number;
  subtotal: number;
};

export type SellerShopOrder = {
  id: string;
  customer_name: string;
  status: "payment_pending" | "commande" | "preparation" | "livraison" | "recu" | "CLAIMED";
  payment_mode: "nita" | "amana";
  payment_status: "pending" | "paid";
  delivery_type: "standard" | "express_niamey";
  tracking_code: string | null;
  total_amount: number;
  currency: string;
  created_at: string;
  items: SellerShopOrderItem[];
};

export type SellerShopOrderStatus = "commande" | "preparation" | "livraison" | "recu";

export type SellerStorefrontProduct = {
  price_id: string;
  product_id: string;
  name: string;
  brand: string;
  description?: string | null;
  amount: number;
  currency: string;
  is_boosted: boolean;
  main_image_url: string | null;
};

export type SellerStorefront = {
  vendor_id: string;
  vendor_slug: string;
  business_name: string;
  activity_type: SellerActivityType;
  storefront_tier: StorefrontTier;
  city: string | null;
  address: string | null;
  phone: string | null;
  description: string | null;
  logo_url: string | null;
  cover_image_url: string | null;
  opening_hours: string | null;
  whatsapp_contact: string | null;
  contact_email: string | null;
  gallery_images: string[];
  service_offerings: SellerServiceOffering[];
  room_types: HotelRoomType[];
  deposit_payment_method: "nita" | "amana" | null;
  deposit_amount: number | null;
  effective_commission_rate: number;
  effective_service_fee: number;
  accepts_table_reservations: boolean;
  accepts_hotel_bookings: boolean;
  is_verified: boolean;
  products: SellerStorefrontProduct[];
  restaurant_menu: RestaurantMenuItem[];
};

export type HotelBookingRequest = {
  vendor_id: string;
  room_type_id: string;
  guest_name: string;
  guest_phone: string;
  guest_email?: string;
  check_in_date: string;
  check_out_date: string;
  guest_count: number;
  deposit_payment_method: "nita" | "amana";
  transaction_reference?: string;
  special_request?: string;
};

export type HotelBooking = {
  id: string;
  vendor_id: string;
  room_type_id: string;
  room_snapshot: Record<string, unknown>;
  guest_name: string;
  guest_phone: string;
  guest_email: string | null;
  check_in_date: string;
  check_out_date: string;
  guest_count: number;
  deposit_payment_method: "nita" | "amana";
  deposit_amount: number;
  transaction_reference: string | null;
  special_request: string | null;
  status: "pending" | "confirmed" | "cancelled";
  created_at: string;
};

File: backend/app/schemas/seller.py
from __future__ import annotations

from datetime import date, datetime
from typing import Any, Literal

from pydantic import BaseModel, ConfigDict, Field

from app.schemas.restaurant import RestaurantMenuItemResponse


class SellerServiceOfferingSchema(BaseModel):
    model_config = ConfigDict(extra="forbid", str_strip_whitespace=True)

    title: str = Field(min_length=1, max_length=120)
    description: str | None = Field(default=None, max_length=600)
    display_mode: Literal["consult_only", "book_only", "shop"] = "consult_only"


class HotelRoomTypeSchema(BaseModel):
    model_config = ConfigDict(extra="forbid", str_strip_whitespace=True)

    id: str | None = Field(default=None, max_length=64)
    name: str = Field(min_length=1, max_length=120)
    description: str | None = Field(default=None, max_length=1000)
    night_price: float = Field(gt=0)
    capacity: int = Field(default=1, ge=1, le=12)
    amenities: list[str] = Field(default_factory=list, max_length=20)
    photo_urls: list[str] = Field(default_factory=list, max_length=20)
    deposit_amount: float | None = Field(default=None, ge=0)


class SellerProfileRequest(BaseModel):
    model_config = ConfigDict(extra="forbid", str_strip_whitespace=True)

    business_name: str = Field(min_length=2, max_length=140)
    phone: str | None = Field(default=None, max_length=40)
    city: str = Field(default="Niamey", min_length=2, max_length=80)
    address: str | None = Field(default=None, max_length=220)
    activity_type: Literal["shop", "restaurant", "hotel", "enterprise"] = "shop"
    storefront_tier: Literal["basic", "premium"] = "basic"
    description: str | None = Field(default=None, max_length=2000)
    logo_url: str | None = Field(default=None, max_length=1024)
    cover_image_url: str | None = Field(default=None, max_length=1024)
    opening_hours: str | None = Field(default=None, max_length=240)
    whatsapp_contact: str | None = Field(default=None, max_length=40)
    contact_email: str | None = Field(default=None, max_length=320)
    gallery_images: list[str] = Field(default_factory=list, max_length=12)
    service_offerings: list[SellerServiceOfferingSchema] = Field(default_factory=list, max_length=20)
    room_types: list[HotelRoomTypeSchema] = Field(default_factory=list, max_length=24)
    deposit_payment_method: Literal["nita", "amana"] | None = None
    deposit_amount: float | None = Field(default=None, ge=0)
    accepts_table_reservations: bool = False
    accepts_hotel_bookings: bool = False


class SellerProfileResponse(BaseModel):
    model_config = ConfigDict(extra="forbid")

    id: str
    user_id: str
    vendor_id: str
    business_name: str
    phone: str | None
    city: str
    address: str | None
    activity_type: Literal["shop", "restaurant", "hotel", "enterprise"]
    storefront_tier: Literal["basic", "premium"]
    description: str | None
    logo_url: str | None
    cover_image_url: str | None
    opening_hours: str | None
    whatsapp_contact: str | None
    contact_email: str | None
    gallery_images: list[str]
    service_offerings: list[SellerServiceOfferingSchema]
    room_types: list[HotelRoomTypeSchema]
    deposit_payment_method: Literal["nita", "amana"] | None
    deposit_amount: float | None
    commission_rate_override: float | None = None
    service_fee_override: float | None = None
    seller_subscription_fee_override: float | None = None
    onboarding_fee_paid_at: datetime | None = None
    subscription_paid_until: datetime | None = None
    subscription_last_payment_reference: str | None = None
    effective_commission_rate: float
    effective_service_fee: float
    effective_seller_subscription_fee: float
    accepts_table_reservations: bool
    accepts_hotel_bookings: bool
    is_verified: bool
    created_at: datetime


class SellerProductCreateRequest(BaseModel):
    model_config = ConfigDict(extra="forbid", str_strip_whitespace=True)

    name: str = Field(min_length=2, max_length=200)
    brand: str = Field(min_length=1, max_length=120)
    description: str | None = Field(default=None, max_length=2000)
    main_image_url: str | None = Field(default=None, max_length=1024)
    category_id: str | None = Field(default=None, max_length=36)
    amount: float = Field(gt=0)
    currency: str = Field(default="XOF", min_length=3, max_length=3)
    stock_quantity: int = Field(default=0, ge=0)
    is_sponsored: bool = False


class SellerProductCreateResponse(BaseModel):
    model_config = ConfigDict(extra="forbid")

    product_id: str
    price_id: str
    vendor_id: str


class SellerInventoryItemResponse(BaseModel):
    model_config = ConfigDict(extra="forbid")

    price_id: str
    product_id: str
    product_name: str
    brand: str
    description: str | None = None
    main_image_url: str | None = None
    amount: float
    currency: str
    stock_quantity: int
    is_active: bool
    is_boosted: bool
    promo_price: float | None = None
    promo_until: datetime | None = None
    boost_until: datetime | None = None


class SellerInventoryUpdateRequest(BaseModel):
    model_config = ConfigDict(extra="forbid")

    product_name: str | None = Field(default=None, min_length=2, max_length=180)
    amount: float | None = Field(default=None, gt=0)
    stock_quantity: int | None = Field(default=None, ge=0)
    description: str | None = Field(default=None, max_length=2000)
    main_image_url: str | None = Field(default=None, max_length=1024)
    is_active: bool | None = None
    promo_amount: float | None = Field(default=None, gt=0)
    boost_duration_hours: int | None = Field(default=None, ge=24, le=168)


class SellerShopOrderItemResponse(BaseModel):
    model_config = ConfigDict(extra="forbid")

    id: str
    product_id: str
    product_name: str
    quantity: int
    unit_price: float
    subtotal: float


class SellerShopOrderResponse(BaseModel):
    model_config = ConfigDict(extra="forbid")

    id: str
    customer_name: str
    status: Literal["payment_pending", "commande", "preparation", "livraison", "recu", "CLAIMED"]
    payment_mode: Literal["nita", "amana"]
    payment_status: Literal["pending", "paid"]
    delivery_type: Literal["standard", "express_niamey"]
    tracking_code: str | None
    total_amount: float
    currency: str
    created_at: datetime
    items: list[SellerShopOrderItemResponse]


class SellerShopOrderStatusUpdateRequest(BaseModel):
    model_config = ConfigDict(extra="forbid")

    status: Literal["commande", "preparation", "livraison", "recu"]


class SellerSubscriptionStatusResponse(BaseModel):
    model_config = ConfigDict(extra="forbid")

    has_seller_profile: bool
    onboarding_fee_paid: bool
    onboarding_fee_paid_at: datetime | None = None
    subscription_paid_until: datetime | None = None
    subscription_active: bool
    monthly_fee: float
    onboarding_fee: float
    amount_due_now: float
    currency: str = "XOF"
    has_pending_payment_request: bool = False


class SellerSubscriptionPayRequest(BaseModel):
    model_config = ConfigDict(extra="forbid", str_strip_whitespace=True)

    payment_mode: Literal["nita", "amana"]
    transaction_reference: str | None = Field(default=None, max_length=180)
    months: int = Field(default=1, ge=1, le=12)


class SellerSubscriptionPaymentRequestResponse(BaseModel):
    model_config = ConfigDict(extra="forbid")

    id: str
    seller_profile_id: str
    seller_user_id: str
    payment_mode: Literal["nita", "amana"]
    transaction_reference: str
    months: int
    amount_claimed: float
    status: Literal["pending", "approved", "rejected"]
    admin_note: str | None = None
    reviewed_by_user_id: str | None = None
    reviewed_at: datetime | None = None
    created_at: datetime


class SellerStorefrontProductResponse(BaseModel):
    model_config = ConfigDict(extra="forbid")

    price_id: str
    product_id: str
    name: str
    brand: str
    description: str | None = None
    amount: float
    currency: str
    is_boosted: bool
    main_image_url: str | None


class SellerStorefrontResponse(BaseModel):
    model_config = ConfigDict(extra="forbid")

    vendor_id: str
    vendor_slug: str
    business_name: str
    activity_type: Literal["shop", "restaurant", "hotel", "enterprise"]
    storefront_tier: Literal["basic", "premium"]
    city: str | None
    address: str | None
    phone: str | None
    description: str | None
    logo_url: str | None
    cover_image_url: str | None
    opening_hours: str | None
    whatsapp_contact: str | None
    contact_email: str | None
    gallery_images: list[str]
    service_offerings: list[SellerServiceOfferingSchema]
    room_types: list[HotelRoomTypeSchema]
    deposit_payment_method: Literal["nita", "amana"] | None
    deposit_amount: float | None
    effective_commission_rate: float
    effective_service_fee: float
    accepts_table_reservations: bool
    accepts_hotel_bookings: bool
    is_verified: bool
    products: list[SellerStorefrontProductResponse]
    restaurant_menu: list[RestaurantMenuItemResponse]


class HotelBookingCreateRequest(BaseModel):
    model_config = ConfigDict(extra="forbid", str_strip_whitespace=True)

    vendor_id: str = Field(min_length=1, max_length=36)
    room_type_id: str = Field(min_length=1, max_length=64)
    guest_name: str = Field(min_length=2, max_length=120)
    guest_phone: str = Field(min_length=6, max_length=40)
    guest_email: str | None = Field(default=None, max_length=320)
    check_in_date: date
    check_out_date: date
    guest_count: int = Field(default=1, ge=1, le=12)
    deposit_payment_method: Literal["nita", "amana"]
    transaction_reference: str | None = Field(default=None, max_length=180)
    special_request: str | None = Field(default=None, max_length=1000)


class HotelBookingStatusUpdateRequest(BaseModel):
    model_config = ConfigDict(extra="forbid")

    status: Literal["pending", "confirmed", "cancelled"]


class HotelBookingResponse(BaseModel):
    model_config = ConfigDict(extra="forbid")

    id: str
    vendor_id: str
    room_type_id: str
    room_snapshot: dict[str, Any]
    guest_name: str
    guest_phone: str
    guest_email: str | None
    check_in_date: date
    check_out_date: date
    guest_count: int
    deposit_payment_method: Literal["nita", "amana"]
    deposit_amount: float
    transaction_reference: str | None
    special_request: str | None
    status: Literal["pending", "confirmed", "cancelled"]
    created_at: datetime
